### ISSUE TITLE
Deleting feature

### ISSUE THREAD

**Issue by wrousseau** [brikteknologier/seraph-model#71]
Is there a reason why you chose not to include a deleting feature for nodes ? If not, would you be interested if I gave the implementation a shot ?

**Comment by IcanDivideBy0** [comment]
Just want to mention that if such a feature appears in a future release, it would be nice to have `beforeDelete` and `afterDelete` events.

**Comment by RubenCordeiro** [comment]
Yes, this would be a great feature to have, specially with cascading deletes. Once I get some spare time I'll try to tackle the beast.

**Comment by andyrichardson** [comment]
Any update on the addition of this feature? I think it is well overdue. 

I would be happy to work on adding it assuming the creators are happy to include such a feature.

**Comment by jonpacker** [comment]
Yep, definitely happy to include it.

**Comment by andyrichardson** [comment]
Apologies, I underestimated the complexity of this project. I assumed all database functions would run through the seraph library but that doesn't seem to be the case. 

I'll leave this task to someone a little more experienced.

**Comment by jonpacker** [comment]
No worries. I think the major barrier to implementing a delete function here is that people would expect it to work with compositions... that's not trivial, and it seems to me like it would be unpredictable.

A good alternative for now is using force deletes (remove relationships as well as the node) in seraph, keeping track of composed orphans, and manually either deleting them or keeping them.

**Comment by andyrichardson** [comment]
If it helps, the way I would have expected it to work would be as follows.

## Functions

**Delete node**
   
    // 'idOrObject' = select node (must be of model type/label)
    // 'force' = boolean to remove all relationships
    model.delete(idOrObject, [force], callback);

**Delete relationship/s only**

    // 'idOrObject' = select/match initial node (must be of model type/label)
    // 'relLabels' = array with relationship labels to be deleted
    // 'relNodeIdOrObject' = target related node (not necessarily same model type/label
    model.deleteRelationship(idOrObject, [relLabels,] [relNodeIdOrObject,] callback);

## Usage Examples

_Assume 'user' and 'session' are models_

**Deleting a single user session**

    session.delete({sessionId: 40}, true, callback);

or

    // Doesn't delete nodes - only relationship/s
    user.deleteRelationship({}, ['has_session'], {sessionId: 40}, callback)

or

    // Doesn't delete nodes - only relationship/s
    session.deleteRelationship({sessionId: 40}, ['has_session'], callback);

**Deleting all user sessions**

    // Doesn't delete nodes - only relationship/s
    user.deleteRelationship({username: andy}, ['has_session'], callback)

or

    session.deleteRelationship({}, ['has_session'], {username: andy}, callback)

**Deleting all user relationships**

    // Doesn't delete nodes - only relationship/s
    user.deleteRelationship({username: andy), [ ], {}, callback)


## Notes

Relationship 'nodes' would need their own model else the following would not be possible.

    MATCH (a:{username:andy})-[r:has_session {type: admin}]-(b) DELETE r

Deleting nodes by relation wouldn't be possible without an extra parameter.

    // 'removeAssoc' = remove associated nodes
    model.deleteRelationship(idOrObject, [removeAssoc,] [relLabels,] [relNodeIdOrObject,] callback)
